src/hashset.c

```diff
@@ -0,0 +1,1928 @@
+    assert(last->presence == 0 || __builtin_popcount(last->presence) == 1);
+    bucketConvertToUnchained(before_last);
+    if (last->presence != 0) {
+        /* Move the last remaining element to the new last position in the
+         * before-last bucket. */
+        int pos_in_last = __builtin_ctz(last->presence);
+        moveElement(before_last, ELEMENTS_PER_BUCKET - 1, last, pos_in_last);
+    }
+    zfree(last);
+    if (s->type->trackMemUsage) s->type->trackMemUsage(s, -sizeof(bucket));
+    s->child_buckets[table_index]--;
+}
+
+/* After removing an element in a bucket with children, we can fill the hole
+ * with an element from the end of the bucket chain and potentially free the
+ * last bucket in the chain. */
+static void fillBucketHole(hashset *s, bucket *b, int pos_in_bucket, int table_index) {
+    assert(b->chained && !isPositionFilled(b, pos_in_bucket));
+    /* Find the last bucket */
+    bucket *before_last = b;
+    bucket *last = bucketNext(b);
+    while (last->chained) {
+        before_last = last;
+        last = bucketNext(last);
+    }
+    /* Unless the last bucket is empty, find an element in the last bucket and
+     * move it to the hole in b. */
+    if (last->presence != 0) {
+        int pos_in_last = __builtin_ctz(last->presence);
+        assert(pos_in_last < ELEMENTS_PER_BUCKET && isPositionFilled(last, pos_in_last));
+        moveElement(b, pos_in_bucket, last, pos_in_last);
+    }
+    /* Free the last bucket if it becomes empty. */
+    if (last->presence == 0 || __builtin_popcount(last->presence) == 1) {
+        pruneLastBucket(s, before_last, last, table_index);
+    }
+}
+
+/* When elements are deleted while rehashing is paused, they leave empty holes
+ * in the buckets. This functions attempts to fill the holes by moving elements
+ * from the end of the bucket chain to fill the holes and free any empty
+ * buckets in the end of the chain. */
+static void compactBucketChain(hashset *s, size_t bucket_index, int table_index) {
+    bucket *b = &s->tables[table_index][bucket_index];
+    while (b->chained) {
+        bucket *next = bucketNext(b);
+        if (next->chained && next->presence == 0) {
+            /* Empty bucket in the middle of the chain. Remove it from the chain. */
+            bucket *next_next = bucketNext(next);
+            b->elements[ELEMENTS_PER_BUCKET - 1] = next_next;
+            zfree(next);
+            if (s->type->trackMemUsage) s->type->trackMemUsage(s, -sizeof(bucket));
+            s->child_buckets[table_index]--;
+            continue;
+        }
+
+        if (!next->chained && (next->presence == 0 || __builtin_popcount(next->presence) == 1)) {
+            /* Next is the last bucket and it's empty or has only one element.
+             * Delete it and turn b into an "unchained" bucket. */
+            pruneLastBucket(s, b, next, table_index);
+            return;
+        }
+
+        if (__builtin_popcount(b->presence) == ELEMENTS_PER_BUCKET - 1) {
```

Comment:
I don't understand this condition. If the number of positions filled is `ELEMENTS_PER_BUCKET - 1`, then there are no holes to be filled.
Would the correct condition be:
`__builtin_popcount(b->presence) < ELEMENTS_PER_BUCKET - 1` ?

Reply:
You're right! Thanks

```suggestion
        if (__builtin_popcount(b->presence) < ELEMENTS_PER_BUCKET - 1) {
```

Reply:
Thanks for your review! This is what's happening when I'm too lazy to write tests. I added a unit test for this now. Also added some testing for memory usage and discovered a bug there too.